controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -256,10 +254,25 @@ else if (completedAt.isPresent()) { // Step not complete, because some jobs rema
                               .flatMap(List::stream);
     }
 
+    /**
+     * Returns the instant when the given application's current change was completed for the given job.
+     *
+     * Any job is complete if its current change was already successful on that job.
+     * A production job is also considered complete if its current change is strictly dominated by what
+     * is already deployed in its zone, i.e., no parts of the change are upgrades, and at least one
+     * part is a downgrade, regardless of the status of the job.
+     */
     private Optional<Instant> completedAt(Application application, JobType jobType) {
-        return jobType.isProduction()
-                ? changeCompletedAt(application, jobType)
-                : application.deploymentJobs().successAt(application.change(), jobType);
+        Optional<Instant> lastSuccess = application.deploymentJobs().successAt(application.change(), jobType);
+        if (lastSuccess.isPresent() || ! jobType.isProduction())
+            return lastSuccess;
+
+        Deployment deployment = application.deployments().get(jobType.zone(controller.system()).get());
+        return Optional.ofNullable(deployment).map(Deployment::at)
+                       .filter(ignored ->    ! (   application.change().upgrades(deployment.version())
+                                                || application.change().upgrades(deployment.applicationVersion()))
+                                          &&   (   application.change().downgrades(deployment.version())
+                                                || application.change().downgrades(deployment.applicationVersion())));
```

Comment:
I think the excessive formatting makes this harder to read when the operators are "far away" from the references.